File: DungeonDebt/Assets/Scripts/Data/EnemyDefinition.cs
using System.Collections.Generic;
using System.Collections.ObjectModel;

public class EnemyDefinition
{
    public EnemyDefinition(
        string id,
        string displayName,
        int attack,
        int health,
        EnemyEffectId effectId,
        string effectDescription,
        params CombatStatusId[] startingStatuses)
        : this(
            id,
            displayName,
            attack,
            health,
            effectId,
            effectDescription,
            startingStatuses,
            null)
    {
    }

    public EnemyDefinition(
        string id,
        string displayName,
        int attack,
        int health,
        EnemyEffectId effectId,
        string effectDescription,
        IReadOnlyList<CombatStatusId> startingStatuses,
        IReadOnlyList<CombatStatusId> attackStatuses)
    {
        Id = id;
        DisplayName = displayName;
        Attack = attack;
        Health = health;
        EffectId = effectId;
        EffectDescription = effectDescription;

        StartingStatuses = BuildStatusList(startingStatuses);
        AttackStatuses = BuildStatusList(attackStatuses);
    }

    public string Id { get; }
    public string DisplayName { get; }
    public int Attack { get; }
    public int Health { get; }
    public EnemyEffectId EffectId { get; }
    public string EffectDescription { get; }
    public IReadOnlyList<CombatStatusId> StartingStatuses { get; }
    public IReadOnlyList<CombatStatusId> AttackStatuses { get; }

    private static IReadOnlyList<CombatStatusId> BuildStatusList(IReadOnlyList<CombatStatusId> source)
    {
        List<CombatStatusId> statuses = new List<CombatStatusId>();
        if (source != null)
        {
            for (int i = 0; i < source.Count; i++)
            {
                if (source[i] != CombatStatusId.None)
                {
                    statuses.Add(source[i]);
                }
            }
        }

        return new ReadOnlyCollection<CombatStatusId>(statuses);
    }
}

File: web/src/data/EnemyDefinition.js
// Ported from DungeonDebt/Assets/Scripts/Data/EnemyDefinition.cs
// Immutable enemy template. startingStatuses/attackStatuses are arrays of
// CombatStatusId, filtered to drop None entries.
import { CombatStatusId } from "./enums.js";

function buildStatusList(source) {
  const statuses = [];
  if (source) {
    for (const s of source) {
      if (s !== CombatStatusId.None) statuses.push(s);
    }
  }
  return Object.freeze(statuses);
}

export class EnemyDefinition {
  constructor(id, displayName, attack, health, effectId, effectDescription, startingStatuses = null, attackStatuses = null) {
    this.id = id;
    this.displayName = displayName;
    this.attack = attack;
    this.health = health;
    this.effectId = effectId;
    this.effectDescription = effectDescription;
    this.startingStatuses = buildStatusList(startingStatuses);
    this.attackStatuses = buildStatusList(attackStatuses);
    Object.freeze(this);
  }
}
